src/python/WMCore/WMSpec/WMWorkload.py

```diff
@@ -68,6 +73,52 @@ class WMWorkloadHelper(PersistencyHelper):
 
     def __init__(self, wmWorkload=None):
         self.data = wmWorkload
+        self.settersMap = {}
+
+    def updateWorkloadArgs(self, reqArgs):
+        """
+        Method to take a dictionary of arguments of the type:
+        {reqArg1: value,
+         reqArg2: value,
+         ...}
+        and update the workload by a predefined map of reqArg to setter methods.
+        :param reqArgs: A Dictionary of request arguments to be updated
+        :return:        Nothing, Raises an error of type WMWorkloadException if
+                        fails to apply the proper setter method
+        """
+        # NOTE: So far we support only a single argument setter methods, like
+        #       setSiteWhitelist or setPriority. This may change in the future,
+        #       but it will require a change in the logic of how we validate and
+        #       call the proper setter methods bellow.
+
+        # populate the current instance settersMap
+        self.settersMap['RequestPriority'] = setterTuple('RequestPriority', self.setPriority, inspect.signature(self.setPriority))
```

Comment:
Given that you are already checking the parameter name and associating it to the actual function (plus inspecting its signature), isn't it more clear to simply call the actual method according to the parameter?

Reply:
Yes, I understand it. What I am saying is that creating a map and passing a pointer to a function does not look as friendly as I classic if-elif dealing with the 3 parameters.

Reply:
creating a map will gives the option of following the same pattern for adding as many setter calls (wfo parameter setter) as one may wish, by providing a generic method of validating and assembling the calls and not changing the logic for the rest of the function (i.e. not forking with hundreds of elifs, and getting into the trap of code growth by number of custom usecases). Here, the addition of a new setter is just a matter of adding yet another line in the map, following the same pattern as the rest in the map i.e.:
```
a named tuple of: ('reqArg', 'setterFunc', 'setterSignature') 
```
So pretty universal way I'd say. And perfectly serves the exact purpose of what is needed here.

Reply:
I understand the rationale for implementing something generic, but right now I am not aware of any other spec parameters that will have to be changed, so it's definitely not hundreds of future elifs.

In addition, I am very much adept of the KISS principle, whenever possible and fit. To me, this overly complex code (personally I don't know any of those inspect methods) can easily lead to a bug, or a buggy development.

Reply:
This is exactly to prevent a bug ... not to cause it.  The inspect method is used only to fetch the signature of the underlying function/method  which is to be used for the setter) and upon that simply validate if the number of arguments passed during the call to the function matches the signature and if not it does not trigger the execution, but signals to the caller accordingly.. That's all. 

The difference between this and an `elif` construct is that in the `elif`  you will have to repeat the same lines of call sequence in every section to the call the exact method based on the request parametr, you will have to hardcode the call without validating during the call if the parameters passed match the current signature and if tomorrow someone changes the signature of the setter method but forgets to change the way how we call it here... voila there is your bug, not the other way around.

Reply:
Well, I actually don't think we need to pre-validate the number of arguments we are passing to a method, because this is already implicit and part of the method execution.

We have a similar method in this very same module that does all of this (used for workflow assignment) and it does not pre-validate number or arguments and/or encapsulate things in try/except, see:
https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/WMSpec/WMWorkload.py#L1999

I still thing we are overdoing things with this implementation. I could be though that I am not well educated or evolving with language, so I would appreciate another opinion on this PR (by no means I intend to push responsibility to others!). @vkuznet @khurtado

Reply:
I guess Andrea wanted to give this answer https://github.com/dmwm/WMCore/pull/12099/files#r1799217523 to this thread instead.

I understand that there are many ways to implement this, either with more or less intellectual knowledge.
So let me ask you this explicitly, @anpicci , do you really think the current implementation is more clear, easier and more maintainable than something like (I don't even think we need try/except here...):
```
        try:
            if 'RequestPriority' in reqArgs:
                self.setPriority(reqArgs['RequestPriority'])
            if 'SiteBlacklist' in reqArgs:
                self.setSiteBlacklist(reqArgs['SiteBlacklist'])
            if 'SiteWhitelist' in reqArgs:
                self.setSiteWhitelist(reqArgs['SiteWhitelist'])
        except Exception as ex:
            msg = f"Failure to update workload parameters. Details: {str(ex)}"
            raise WMWorkloadException(msg) from None
```
?

Reply:
@amaltaro , @anpicci , @todor-ivanov 

Given that Python (or any programming language) offers multiple ways to achieve the same outcome, evaluating the best approach can often be subjective, depending on factors such as experience, proficiency, or personal style. Consequently, these discussions can become matters of personal preference.

For example, choosing between a full for-loop or list comprehension is a matter of choice. While we can debate which one might be "better" based on various criteria, both serve the same functional purpose. In this instance, I feel the this particular conversation is similar, and rather than debating stylistic choices, the focus should be on whether the code achieves its intended purpose. Otherwise, we risk undermining the role of both developer and reviewer by favoring one approach over another, which I find counterproductive.

If there are concerns about correctness or functionality, I would recommend incorporating unit tests for validation instead of continuing to focus on the specific implementation style.

Reply:
This is not coding style and personal preferences, but I see it as code readability and sustainability.

Anyhow, thank you for sharing your perspective on this matter. I won't bug people anymore on this.

Reply:
Alan, 

This: 
```
        try:
            if 'RequestPriority' in reqArgs:
                self.setPriority(reqArgs['RequestPriority'])
            if 'SiteBlacklist' in reqArgs:
                self.setSiteBlacklist(reqArgs['SiteBlacklist'])
            if 'SiteWhitelist' in reqArgs:
                self.setSiteWhitelist(reqArgs['SiteWhitelist'])
        except Exception as ex:
            msg = f"Failure to update workload parameters. Details: {str(ex)}"
            raise WMWorkloadException(msg) from None
```
is simply  not equal to this:
```
        for reqArg, argValue in reqArgs.items():
            if not self.settersMap.get(reqArg, None):
                msg = f"Unsupported or missing setter method for updating reqArg: {reqArg}."
                raise WMWorkloadException(msg)
            try:
                self.settersMap[reqArg].setterSignature.bind(argValue)
            except TypeError as ex:
                msg = f"Setter's method signature does not match the method calls we currently support: Error: req{str(ex)}"
                raise WMWorkloadException(msg) from None

        # Now go through the reqArg again and call every setter method according to the map
        for reqArg, argValue in reqArgs.items():
            try:
                self.settersMap[reqArg].setterFunc(argValue)
            except Exception as ex:
                currFrame = inspect.currentframe()
                argsInfo = inspect.getargvalues(currFrame)
                argVals = {arg: argsInfo.locals.get(arg) for arg in argsInfo.args}
                msg = f"Failure while calling setter method {self.settersMap[reqArg].setterFunc.__name__} "
                msg += f"With arguments: {argVals}"
                msg += f"Full exception string: {str(ex)}"
                raise WMWorkloadException(msg) from None
```

At the end, in those both implementations, if the call to the setter method is wrong they will both fail the call, but the later is much more descriptive. It recognizes (at least  3) different possible conditions of failure and forwards the proper message for later debugging, while the former just fails the call and masks all the relevant information for the one who is to chase eventual problems. And that  was the exact reason why I went for this implementation. I know you'd always prefer the shorter version, but sometimes it hides/masks valuable information.

Reply:
Something more, since  this is supposed to be a method triggered by a user's call rather than an operation triggered by the internals of our system, the later method inspects the stack and gives you the values of all parameters in the call. Meaning it shows you the exact (eventually) user mistake.

Reply:
Talking about inspection and validation. Validation of the user parameters is not supposed to happen at this layer (at the setter call). Validation should be performed in the Service/Request class (which is likely calling the Validation utils).

With that said, I just noticed that we are not validating the site lists, as we usually do for workflow assignment.
So we should add that validation in here as well to protect the system and us from unneded debugging. For reference, this is how we define the site lists and the validation function:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1147